subprojects/base-services/src/main/java/org/gradle/internal/SystemProperties.java

```diff
@@ -133,19 +123,29 @@ public File getJavaHomeDir() {
      * @param factory Instance created by the Factory implementation
      */
     public <T> T withSystemProperty(String propertyName, String value, Factory<T> factory) {
-        lock.lock();
-        String originalValue = System.getProperty(propertyName);
-        System.setProperty(propertyName, value);
+        synchronized (System.getProperties()) {
+            String originalValue = System.getProperty(propertyName);
```

Comment:
I think there's another race here that would be hard to prevent. I don't think it's new:

t1:  synchronized(getProperties())
t2: setProperties(newProperties)
t1: getProperty(...) // not the original properties now and factory will be vulnerable to CME again

or in a similar sort of way:

t1:  synchronized(getProperties())
t1: getProperty(...) // getting from original properties
t1: setProperty(...) // setting to original properties
t2: setProperties(newProperties)
t1: factory.create() // sees a different set of properties
t1: finally {} // resets/clears our property on the "wrong" properties

I don't think we need to do anything about this until it becomes a problem.

Reply:
Oh dear, why would they even allow completely overwriting that field? I agree, this is hard to prevent.